### ISSUE TITLE
Missing documentation on how to implement custom flow_from_X

### ISSUE THREAD

**Issue by mristin** [keras-team/keras-preprocessing#252]
Hi keras-team!
First of all, thanks for all the great work!

I would like to create a PR to extend the documentation on how to implement a custom `ImageDataGenerator.flow_from_...` functionality. 

At the moment, it is very unclear how to achieve reading custom images and labels as soon as `flow_from_directory` or `flow_from_dataframe` do not match the use case.  For example, imagine the labels are embedded in file names and split in different train/val/test directories. The page https://keras.io/preprocessing/image/ does not document `DirectoryIterator` at all and provides no hint how such functionality should be implemented.

Before I extend the docs, could you please tell me what you assume the canonical approach would be? Implementing a custom class akin to `DirectoryIterator`? Shouldn't `DirectoryIterator` also be documented on https://keras.io/preprocessing/image/?

Thanks for the pointers!

**Comment by mristin** [comment]
Here are a couple of additional cases: images and labels from a zip archive, downloaded on-the-fly and labels in a CSV file.
